### PR TITLE
Pick plan tutorial

### PR DESCRIPTION
**Pull request by Ridhwanluthra**

### CODE REVIEW

**Comment by mlautman** [comment]
Include descriptions with your PR's

File: doc/pick_place/pick_place_tutorial.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
+Pick and Place Tutorial
+============================
+
+In MoveIt!, grasping is done using the move group interface. In order to grasp an object we need to create ``moveit_msgs::Grasp`` msg which will allow defining the various poses and postures involved in a grasping operation.
+Watch this video to see the output of this tutorial:
+
+.. raw:: html
+
+    <div style="position: relative; padding-bottom: 5%; height: 0; overflow: hidden; max-width: 100%; height: auto;">
```
Comment (mlautman) [inline doc/pick_place/pick_place_tutorial.rst]: Yes!!!!

File: doc/pick_place/pick_place_tutorial.rst [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+Running The Demo
+----------------
+Open two shells. In the first shell start RViz and wait for everything to finish loading: ::
+
+    roslaunch panda_moveit_config demo.launch
+
+In the second shell run the pick and place tutorial: ::
+    
+    rosrun moveit_tutorials pick_place_tutotrial
```
Comment (mlautman) [inline doc/pick_place/pick_place_tutorial.rst]: You need a roslaunch for this

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 3]
```diff
@@ -1,244 +1,215 @@
-/*********************************************************************
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: Why did you delete this?

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 4]
```diff
@@ -1,244 +1,215 @@
...
-  co.operation = moveit_msgs::CollisionObject::ADD;
-  co.primitives[0].dimensions[shape_msgs::SolidPrimitive::BOX_X] = 0.15;
-  co.primitives[0].dimensions[shape_msgs::SolidPrimitive::BOX_Y] = 0.1;
-  co.primitives[0].dimensions[shape_msgs::SolidPrimitive::BOX_Z] = 0.3;
-
-  co.primitive_poses[0].position.x = 0.6;
-  co.primitive_poses[0].position.y = -0.7;
-  co.primitive_poses[0].position.z = 0.5;
-  pub_co.publish(co);
+  // BEGIN_SUB_TUTORIAL table1
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: This should be in it's own method not main

File: doc/pick_place/pick_place_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
+Pick and Place Tutorial
+============================
+
+In MoveIt!, grasping is done using the move group interface. In order to grasp an object we need to create ``moveit_msgs::Grasp`` msg which will allow defining the various poses and postures involved in a grasping operation.
```
Comment (davetcoleman) [inline doc/pick_place/pick_place_tutorial.rst]: move group -> MoveGroup

File: doc/pick_place/pick_place_tutorial.rst [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+Watch this video to see the output of this tutorial:
+
+.. raw:: html
+
+    <div style="position: relative; padding-bottom: 5%; height: 0; overflow: hidden; max-width: 100%; height: auto;">
+        <iframe width="700px" height="400px" src="https://www.youtube.com/embed/cHxkQwSyrgY?rel=0" frameborder="0" allow="autoplay; encrypted-media" allowfullscreen></iframe>
+    </div>
```
Comment (davetcoleman) [inline doc/pick_place/pick_place_tutorial.rst]: Nice!

File: doc/pick_place/pick_place_tutorial.rst [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+Getting Started
+---------------
+If you haven't already done so, make sure you've completed the steps in `Getting Started <../getting_started/getting_started.html>`_.
+
+Running The Demo
+----------------
+Open two shells. In the first shell start RViz and wait for everything to finish loading: ::
```
Comment (davetcoleman) [inline doc/pick_place/pick_place_tutorial.rst]: shells -> terminals?

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 8]
```diff
@@ -1,244 +1,215 @@
...
+  // Define the pose of the table.
+  collision_objects[0].primitive_poses.resize(1);
+  collision_objects[0].primitive_poses[0].position.x = 0.5;
+  collision_objects[0].primitive_poses[0].position.y = 0;
+  collision_objects[0].primitive_poses[0].position.z = 0.2;
+  // END_SUB_TUTORIAL
+
+  collision_objects[0].operation = collision_objects[0].ADD;
+
+  // BEGIN_SUB_TUTORIAL table2
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: Separate  method

File: index.rst [hunk 9]
```diff
@@ -42,6 +42,7 @@ Building more complex applications with MoveIt! often requires developers to dig
    doc/visualizing_collisions/visualizing_collisions_tutorial
    doc/time_parameterization/time_parameterization_tutorial
    doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial
+   doc/pick_place/pick_place_tutorial
```
Comment (mlautman) [inline index.rst]: This makes me really happy :)

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 10]
```diff
@@ -1,244 +1,215 @@
...
- *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
- *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
- *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
- *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
- *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
- *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
- *  POSSIBILITY OF SUCH DAMAGE.
- *********************************************************************/
-
-/* Author: Ioan Sucan */
```
Comment (davetcoleman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: You must keep the BSD license! Add your name next to Ioan

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 11]
```diff
@@ -1,244 +1,215 @@
...
+
+  // BEGIN_SUB_TUTORIAL pick3
+  // Set support surface as table1.
+  move_group.setSupportSurfaceName("table1");
+  // Call pick to pick up the object using the grasps given
+  move_group.pick("object", grasps);
+  // END_SUB_TUTORIAL
 }
 
 void place(moveit::planning_interface::MoveGroupInterface& group)
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: When I run locally place fails with "All supplied place locations failed. Retrying last location in verbose mode."

File: doc/pick_place/pick_place_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+In the second shell run the pick and place tutorial: ::
+    
+    rosrun moveit_tutorials pick_place_tutotrial
+
+You should see something similar to the video at the beginning of this tutorial.
+
+Understanding ``moveit_msgs::Grasp``
+------------------------------------
+For complete documentation refer to `this. <http://docs.ros.org/kinetic/api/moveit_msgs/html/msg/Grasp.html>`_
```
Comment (mlautman) [inline doc/pick_place/pick_place_tutorial.rst]: Replace `this` with `moveit_msgs/Grasp.msg`

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 13]
```diff
@@ -1,244 +1,215 @@
...
-
-  grasps.push_back(g);
-  group.setSupportSurfaceName("table");
-  group.pick("part", grasps);
+  grasps.resize(1);
+
+  // Setting the grasp pose
+  grasps[0].grasp_pose.header.frame_id = "panda_link0";
+  grasps[0].grasp_pose.pose.orientation =
+      tf::createQuaternionMsgFromRollPitchYaw(-1.5707963267948966, -0.7853981633974483, -1.5707963267948966);
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: use `M_PI` from `math.h`

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 14]
```diff
@@ -1,244 +1,215 @@
...
+  // This is the pose of panda_link8.
+  // From panda_link8 to the palm of the eef the distance is 0.058, the cube starts 0.01 before 5.0 (half of the length
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: Great comments!

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 15]
```diff
@@ -1,244 +1,215 @@
...
- *  POSSIBILITY OF SUCH DAMAGE.
- *********************************************************************/
-
-/* Author: Ioan Sucan */
-
+// ROS
 #include <ros/ros.h>
 
 // MoveIt!
-#include <moveit/planning_scene_monitor/planning_scene_monitor.h>
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: Please explain this decision

**Comment by mlautman** [comment]
@Ridhwanluthra Ping us when this is ready for another review

File: doc/pick_place/pick_place_tutorial.rst (hunk 2 above)
Comment (Ridhwanluthra) [reply doc/pick_place/pick_place_tutorial.rst]: i had initially created a launch but the problem is that we need to wait for everything to load before running this node.

File: doc/pick_place/src/pick_place_tutorial.cpp (hunk 11 above)
Comment (Ridhwanluthra) [reply doc/pick_place/src/pick_place_tutorial.cpp]: It happens here also rarely. I am guessing its because there is only one grasp given and it only tries to plan once.

**Comment by Ridhwanluthra** [comment]
@mlautman could you please review this again.

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 16]
```diff
@@ -45,37 +82,46 @@ void pick(moveit::planning_interface::MoveGroupInterface& move_group)
...
-  // This is the pose of panda_link8.
-  // From panda_link8 to the palm of the eef the distance is 0.058, the cube starts 0.01 before 5.0 (half of the length
-  // of the cube).
-  // Therfore, the posotion for panda_link8 = 5 - (length of cube/2 - distance b/w panda_link8 and palm of eef - some
-  // extra padding
+      tf::createQuaternionMsgFromRollPitchYaw(-M_PI / 2, -M_PI / 4, -M_PI / 2);
+  /* This is the pose of panda_link8. */
+  /* From panda_link8 to the palm of the eef the distance is 0.058, the cube starts 0.01 before 5.0 (half of the length */
+  /* of the cube). */
+  /* Therfore, the posotion for panda_link8 = 5 - (length of cube/2 - distance b/w panda_link8 and palm of eef - some */
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: position

You should use a spell checker to catch typos like this.

File: doc/pick_place/src/pick_place_tutorial.cpp (hunk 11 above)
Comment (mlautman) [reply doc/pick_place/src/pick_place_tutorial.cpp]: Add a comment that this is a known issue with a TODO to fix.

File: doc/pick_place/src/pick_place_tutorial.cpp [hunk 17]
```diff
@@ -305,4 +307,4 @@ int main(int argc, char** argv)
 // Place Pipeline
 // ^^^^^^^^^^^^^^
 // CALL_SUB_TUTORIAL place
-// END_TUTORIAL
\ No newline at end of file
+// END_TUTORIAL
```
Comment (mlautman) [inline doc/pick_place/src/pick_place_tutorial.cpp]: Great job!

File: doc/pick_place/pick_place_tutorial.rst [hunk 18]
```diff
@@ -43,4 +43,8 @@ The Entire Code
 ---------------
 The entire code can be seen :codedir:`here <pick_place>` in the moveit_tutorials GitHub project.
 
-.. tutorial-formatter:: ./src/pick_place_tutorial.cpp
\ No newline at end of file
+.. |br| raw:: html
```
Comment (mlautman) [inline doc/pick_place/pick_place_tutorial.rst]: Is this just for a line break? Seems like overkill

File: doc/pick_place/src/pick_place_tutorial.cpp (hunk 11 above)
Comment (mlautman) [reply doc/pick_place/src/pick_place_tutorial.cpp]: Would generating multiple grasps fix this?

**Comment by mlautman** [comment]
Congrats @Ridhwanluthra!

File: doc/pick_place/pick_place_tutorial.rst (hunk 18 above)
Comment (Ridhwanluthra) [reply doc/pick_place/pick_place_tutorial.rst]: I couldn't find a better way to do this and it looks real shabby without it.
